relayer/chain/relaychain/connection.go

```diff
@@ -99,5 +101,106 @@ func (co *Connection) GetMMRLeafForBlock(
 		"Proof.LeafCount":                 proofResponse.Proof.LeafCount,
 		"Proof.Items":                     proofItemsHex,
 	}).Info("Generated MMR Proof")
-	return proofResponse
+	return proofResponse, nil
+}
+
+func (co *Connection) GetAllParaheadsWithOwn(blockHash types.Hash, ownParachainId uint32) ([]types.Header, types.Header, error) {
+	none := types.NewOptionU32Empty()
+	encoded, err := types.EncodeToBytes(none)
+	if err != nil {
+		co.log.WithError(err).Error("Error")
+		return nil, types.Header{}, err
+	}
+
+	baseParaHeadsStorageKey, err := types.CreateStorageKey(
+		co.GetMetadata(),
+		"Paras",
+		"Heads", encoded, nil)
+	if err != nil {
+		co.log.WithError(err).Error("Failed to create parachain header storage key")
+		return nil, types.Header{}, err
+	}
+
+	//TODO fix this manual slice.
+	// The above types.CreateStorageKey does not give the same base key as polkadotjs needs for getKeys.
+	// It has some extra bytes.
+	// maybe from the none u32 in golang being wrong, or maybe slightly off CreateStorageKey call? we slice it
+	// here as a hack.
+	actualBaseParaHeadsStorageKey := baseParaHeadsStorageKey[:32]
+	co.log.WithField("actualBaseParaHeadsStorageKey", actualBaseParaHeadsStorageKey.Hex()).Info("actualBaseParaHeadsStorageKey")
+
+	keysResponse, err := co.GetAPI().RPC.State.GetKeys(actualBaseParaHeadsStorageKey, blockHash)
+	if err != nil {
+		co.log.WithError(err).Error("Failed to get all parachain keys")
+		return nil, types.Header{}, err
+	}
+
+	headersResponse, err := co.GetAPI().RPC.State.QueryStorage(keysResponse, blockHash, blockHash)
+	if err != nil {
+		co.log.WithError(err).Error("Failed to get all parachain headers")
+		return nil, types.Header{}, err
+	}
+
+	co.log.Info("Got all parachain headers")
+	var headers []types.Header
+	var ownParachainHeader types.Header
+	for _, headerResponse := range headersResponse {
+		for _, change := range headerResponse.Changes {
+
+			// TODO fix this manual slice with a proper type decode. only the last few bytes are for the ParaId,
+			// not sure what the early ones are for.
+			key := change.StorageKey[40:]
+			var parachainID types.U32
+			if err := types.DecodeFromBytes(key, &parachainID); err != nil {
+				co.log.WithError(err).Error("Failed to decode parachain ID")
+				return nil, types.Header{}, err
+			}
+
+			co.log.WithField("parachainId", parachainID).Info("Decoding header for parachain")
+			var encodableOpaqueHeader types.Bytes
+			if err := types.DecodeFromBytes(change.StorageData, &encodableOpaqueHeader); err != nil {
+				co.log.WithError(err).Error("Failed to decode MMREncodableOpaqueLeaf")
+				return nil, types.Header{}, err
+			}
+
+			var header types.Header
+			if err := types.DecodeFromBytes(encodableOpaqueHeader, &header); err != nil {
```

Comment:
As I said in an earlier comment, this will fail if the para head doesn't conform to the default Substrate header format. Are you planning to fix in a follow-up PR, or can you just skip non-Snowbridge heads here and now?

Reply:
ah i forgot about this, thanks will do in this PR

Reply:
@vgeddes https://github.com/Snowfork/polkadot-ethereum/pull/446/commits/b796afe077c4122542ecc85c44478e261cf94cbd

Reply:
Looks good :+1: